src/main/java/com/limechain/network/protocol/grandpa/GrandpaMessageHandler.java

```diff
@@ -331,10 +336,26 @@ public void handleCatchUpResponse(PeerId peerId,
         setPreVotesAndPvEquivocations(grandpaRound, catchUpResMessage.getPreVotes());
         setPreCommitsAndPcEquivocations(grandpaRound, catchUpResMessage.getPreCommits());
 
-        boolean verified = JustificationVerifier.verify(Justification.fromCatchUpResMessage(catchUpResMessage));
+        CompletableFuture<Boolean> verifiedPreVotesFuture = asyncExecutor.executeAsync(() ->
+                JustificationVerifier.verify(Justification.fromCatchUpResPreVotes(catchUpResMessage)));
 
-        if (!verified) {
-            throw new JustificationVerificationException("Justification could not be verified.");
+        CompletableFuture<Boolean> verifiedPreCommitsFuture = asyncExecutor.executeAsync(() ->
+                JustificationVerifier.verify(Justification.fromCatchUpResPreCommits(catchUpResMessage)));
+
+        // Combines verified of preVotes and preCommits - it is true only if both and verified.
+        CompletableFuture<Boolean> verifiedFuture = verifiedPreVotesFuture.thenCombine(verifiedPreCommitsFuture,
+                (preVotes, preCommits) -> preVotes && preCommits);
+
+        try {
+            boolean verified = verifiedFuture.get();
```

Comment:
```suggestion
            boolean verified = verifiedFuture.join();
```

You can use join() to avoid the try catch here.